File: packages/dotnet/OpenSkp/JsonExport.cs
using System.Collections.Generic;
using System.Linq;

namespace OpenSkp
{
    /// <summary>openskp's canonical JSON export schema, shared with the
    /// Python (<c>to_dict</c>) and TypeScript (<c>toJSON</c>) ports. Every
    /// definition (root included) carries full vertex/edge/face arrays and
    /// counts, plus its raw (pre-bake) <c>instances</c> list; pass a
    /// <see cref="Scene"/> (the result of
    /// <see cref="SkpFile.BuildScene(string, SkpParseOptions)"/>) to also
    /// include the resolved, world-space <c>scene_hierarchy</c>/
    /// <c>mesh_index</c> - omit it for a lighter summary covering just the
    /// raw model.
    ///
    /// The raw per-definition <c>instances</c> list is intentionally flat,
    /// with no <c>layer</c>/<c>properties</c>/<c>children</c> keys: those
    /// are declared on <see cref="Instance"/> but never assigned during
    /// parsing here (same as Python's and Dart's <c>Instance</c>; only C++
    /// actually populates them - see item 17), so encoding them would
    /// present known-dead data as meaningful. The resolved, genuinely
    /// nested tree (with correct layer/properties) is available via
    /// <c>scene_hierarchy</c> instead.</summary>
    public static class JsonExport
    {
        private static Dictionary<string, object?> InstanceToJson(Instance inst) => new Dictionary<string, object?>
        {
            ["name"] = inst.Name,
            ["ref_idx"] = inst.RefIdx,
            ["guid"] = inst.Guid,
            ["matrix"] = inst.Matrix.Cast<object>().ToList(),
        };

        private static Dictionary<string, object?> VertexToJson(Vertex v) => new Dictionary<string, object?>
        {
            ["id"] = v.Id, ["x"] = v.X, ["y"] = v.Y, ["z"] = v.Z,
        };

        private static Dictionary<string, object?> EdgeToJson(Edge e) => new Dictionary<string, object?>
        {
            ["id"] = e.Id, ["v1_id"] = e.V1Id, ["v2_id"] = e.V2Id,
        };

        private static Dictionary<string, object?> FaceToJson(Face f)
        {
            var loops = f.Loops.Select(loop => (object)loop.Select(ce => (object)new Dictionary<string, object?>
            {
                ["edge_id"] = ce.EdgeId,
                ["orientation"] = ce.Orientation,
            }).ToList()).ToList();

            object? normal = f.Normal.HasValue
                ? new List<object> { f.Normal.Value.Nx, f.Normal.Value.Ny, f.Normal.Value.Nz }
                : null;

            return new Dictionary<string, object?>
            {
                ["id"] = f.Id,
                ["loops"] = loops,
                ["normal"] = normal,
            };
        }

        private static Dictionary<string, object?> DefinitionToJson(Definition defn) => new Dictionary<string, object?>
        {
            ["id"] = defn.Id,
            ["guid"] = defn.Guid,
            ["name"] = defn.Name,
            ["vertex_count"] = defn.Vertices.Count,
            ["edge_count"] = defn.Edges.Count,
            ["face_count"] = defn.Faces.Count,
            ["vertices"] = defn.Vertices.Values.Select(v => (object)VertexToJson(v)).ToList(),
            ["edges"] = defn.Edges.Values.Select(e => (object)EdgeToJson(e)).ToList(),
            ["faces"] = defn.Faces.Values.Select(f => (object)FaceToJson(f)).ToList(),
            ["instances"] = defn.Instances.Select(i => (object)InstanceToJson(i)).ToList(),
        };

        private static Dictionary<string, object?> InstanceNodeToJson(InstanceNode node) => new Dictionary<string, object?>
        {
            ["name"] = node.Name,
            ["definition_name"] = node.DefinitionName,
            ["layer"] = node.Layer,
            ["position_mm"] = new List<object> { node.PositionMm.X, node.PositionMm.Y, node.PositionMm.Z },
            ["properties"] = node.Properties.ToDictionary(kv => kv.Key, kv => (object)kv.Value),
            ["children"] = node.Children.Select(c => (object)InstanceNodeToJson(c)).ToList(),
        };

        private static Dictionary<string, object?> MeshMetadataToJson(MeshMetadata m) => new Dictionary<string, object?>
        {
            ["name"] = m.Name,
            ["definition_name"] = m.DefinitionName,
            ["layer"] = m.Layer,
            ["position_mm"] = new List<object> { m.PositionMm.X, m.PositionMm.Y, m.PositionMm.Z },
            ["properties"] = m.Properties.ToDictionary(kv => kv.Key, kv => (object)kv.Value),
            ["path"] = m.Path,
        };

        /// <summary>Convert a parsed <see cref="SkpModel"/> (and optionally a
        /// baked <see cref="Scene"/>) to a JSON-serializable dictionary.
        /// Pass the result to <see cref="MiniJson.Serialize"/> to get a JSON
        /// string.</summary>
        public static Dictionary<string, object?> ToDict(SkpModel model, Scene? scene = null)
        {
            var definitionsObj = new Dictionary<string, object?>();
            foreach (var kv in model.Definitions)
            {
                definitionsObj[kv.Key.ToString()] = DefinitionToJson(kv.Value);
            }

            var layersList = model.Layers.Select(l => (object)new Dictionary<string, object?>
            {
                ["name"] = l.Name,
                ["color"] = new Dictionary<string, object?> { ["r"] = l.ColorR, ["g"] = l.ColorG, ["b"] = l.ColorB },
                ["hidden"] = l.Hidden,
            }).ToList();

            var materialsList = model.Materials.Select(m => (object)new Dictionary<string, object?>
            {
                ["name"] = m.Name,
                ["color"] = new Dictionary<string, object?>
                {
                    ["r"] = m.Color.R, ["g"] = m.Color.G, ["b"] = m.Color.B, ["a"] = m.Color.A,
                },
                ["transparency"] = m.Transparency,
            }).ToList();

            var meshIndexObj = new Dictionary<string, object?>();
            if (scene != null)
            {
                foreach (var kv in scene.MeshIndex)
                {
                    meshIndexObj[kv.Key] = MeshMetadataToJson(kv.Value);
                }
            }

            return new Dictionary<string, object?>
            {
                ["format_version"] = "1.0",
                ["sketchup_version"] = model.Version,
                ["units"] = model.Units,
                ["total_definitions"] = model.Definitions.Count,
                ["total_layers"] = model.Layers.Count,
                ["total_meshes"] = scene != null ? scene.MeshIndex.Count : 0,
                ["root"] = DefinitionToJson(model.Root),
                ["definitions"] = definitionsObj,
                ["layers"] = layersList,
                ["materials"] = materialsList,
                ["mesh_index"] = meshIndexObj,
                ["scene_hierarchy"] = scene != null ? InstanceNodeToJson(scene.SceneHierarchy) : null,
            };
        }
    }
}

File: packages/dart/lib/src/json_export.dart
import 'model.dart';
import 'scene.dart';

/// openskp's canonical JSON export schema, shared with the Python
/// (`to_dict`) and TypeScript (`toJSON`) ports. Every definition (root
/// included) carries full vertex/edge/face arrays and counts, plus its
/// raw (pre-bake) `instances` list; pass [scene] (the result of
/// `SkpFile.buildScene()`) to also include the resolved, world-space
/// `scene_hierarchy`/`mesh_index` - omit it for a lighter summary
/// covering just the raw model.
///
/// The raw per-definition `instances` list is intentionally flat, with
/// no `layer`/`properties`/`children` keys: those are declared on
/// [Instance] but never assigned during parsing here (same as Python's
/// and .NET's `Instance`; only C++ actually populates them - see item
/// 17), so encoding them would present known-dead data as meaningful.
/// The resolved, genuinely nested tree (with correct layer/properties)
/// is available via `scene_hierarchy` instead.
Map<String, dynamic> _instanceToJson(Instance inst) => {
      'name': inst.name,
      'ref_idx': inst.refIdx,
      'guid': inst.guid,
      'matrix': inst.matrix,
    };

Map<String, dynamic> _vertexToJson(Vertex v) => {'id': v.id, 'x': v.x, 'y': v.y, 'z': v.z};

Map<String, dynamic> _edgeToJson(Edge e) => {'id': e.id, 'v1_id': e.v1Id, 'v2_id': e.v2Id};

Map<String, dynamic> _faceToJson(Face f) => {
      'id': f.id,
      'loops': f.loops
          .map((loop) => loop.map((ce) => {'edge_id': ce.$1, 'orientation': ce.$2}).toList())
          .toList(),
      'normal': f.normal == null ? null : [f.normal!.$1, f.normal!.$2, f.normal!.$3],
    };

Map<String, dynamic> _definitionToJson(Definition defn) => {
      'id': defn.id,
      'guid': defn.guid,
      'name': defn.name,
      'vertex_count': defn.vertices.length,
      'edge_count': defn.edges.length,
      'face_count': defn.faces.length,
      'vertices': defn.vertices.values.map(_vertexToJson).toList(),
      'edges': defn.edges.values.map(_edgeToJson).toList(),
      'faces': defn.faces.values.map(_faceToJson).toList(),
      'instances': defn.instances.map(_instanceToJson).toList(),
    };

Map<String, dynamic> _instanceNodeToJson(InstanceNode node) => {
      'name': node.name,
      'definition_name': node.definitionName,
      'layer': node.layer,
      'position_mm': [node.positionMm.$1, node.positionMm.$2, node.positionMm.$3],
      'properties': node.properties,
      'children': node.children.map(_instanceNodeToJson).toList(),
    };

Map<String, dynamic> _meshMetadataToJson(MeshMetadata m) => {
      'name': m.name,
      'definition_name': m.definitionName,
      'layer': m.layer,
      'position_mm': [m.positionMm.$1, m.positionMm.$2, m.positionMm.$3],
      'properties': m.properties,
      'path': m.path,
    };

/// Convert a parsed [SkpModel] (and optionally a baked [Scene]) to a
/// JSON-serializable [Map]. See this file's top-level docs for the
/// schema this matches exactly (Python's `to_dict`, TypeScript's
/// `toJSON`).
Map<String, dynamic> toJson(SkpModel model, [Scene? scene]) {
  final definitionsObj = <String, dynamic>{};
  for (final entry in model.definitions.entries) {
    definitionsObj[entry.key.toString()] = _definitionToJson(entry.value);
  }

  return {
    'format_version': '1.0',
    'sketchup_version': model.version,
    'units': model.units,
    'total_definitions': model.definitions.length,
    'total_layers': model.layers.length,
    'total_meshes': scene != null ? scene.meshIndex.length : 0,
    'root': _definitionToJson(model.root),
    'definitions': definitionsObj,
    'layers': model.layers
        .map((l) => {
              'name': l.name,
              'color': {'r': l.colorR, 'g': l.colorG, 'b': l.colorB},
              'hidden': l.hidden,
            })
        .toList(),
    'materials': model.materials
        .map((m) => {
              'name': m.name,
              'color': {'r': m.color.$1, 'g': m.color.$2, 'b': m.color.$3, 'a': m.color.$4},
              'transparency': m.transparency,
            })
        .toList(),
    'mesh_index': scene != null
        ? {for (final e in scene.meshIndex.entries) e.key: _meshMetadataToJson(e.value)}
        : <String, dynamic>{},
    'scene_hierarchy': scene != null ? _instanceNodeToJson(scene.sceneHierarchy) : null,
  };
}
